docs/compilers/terrapin.md

```diff
@@ -0,0 +1,125 @@
+
+The expectation is that Terrapin will install this as a local tool for a given validation event. That will allow for every validation event to use a different version of `dotnet-roslyn-validator` without having to worry about version conflicts as it would if it were installed as a global tool.
+
+## The Terrapin Service Workflow
+The Terrapin service will validate a DLL by going through the following workflow for every DLL in a NuPkg file:
+
+1. Execute `dotnet-roslyn-manifest-generator` on the DLL + PDB conversation. 
+1. Setup the environment for validation:
+    1. Transfer execution to the appropriate operating system listed in the manifest. All future actions are expected to run on this machine.
+    1. Install the correct .NET Runtime on the target machine (if necessary)
+1. Download the artifacts listed in the artifact manifest file as well as the specified version of `dotnet-roslyn-validator`. 
+1. Execute `dotnet-roslyn-validator` providing the DLL, PDB and the directory where the artifacts were downloaded too.
+
+The Terrapin service will download the artifacts into a directory created by Terrapin. The directory will have two directories to store source files and references respectively: sources and references. The name of the artifacts in those directories will be specified in the manifest file. That will put the burden of avoiding collisions and dealing with file path fixup on the compiler team.
+
+## Artifacts Manifest File
+The goal of `dotnet-roslyn-manifest-generator` is to provide a manifest file to the Terrapin service which lists all of the artifacts necessary for a validation to be completed as well as the environment needed to run the validation. The set of artifacts that are necessary include:
+
+1. The source files used in the compilation. 
+1. The reference DLL used in the compilation. 
+
+At this time the expectation is Terrapin will need to recreate much of the environment used to create the binaries in order to perform provenance validation. This includes having a matching .NET Runtime version and operating system. This is because OS dependencies, like zip compression version, can impact the deterministic output of the compiler. It's possible our experiments in Terrapin will cause us to re-evaluate how these play into determinism but for now the design is to build validation on top of our existing assumptions here. 
+
+To facilitate this the artifact manifest file will include all of the necessary environment information. This will include the OS RID, the .NET Runtime version and whether or a desktop or NET Core runtime was used to run the compiler.
+
+The artifacts manifest file will be a JSON file with the following format:
+
+```json
+{
+    "sources": [
+        { 
+            "uri": "https://github.com/dotnet/roslyn/blob/f5517311e37a3883e09dc6daa5069bbc4f44c882/src/Compilers/Shared/Csc.cs",
```

Comment:
We also need the name of the file as listed in the PDB. That info is needed for `CallerFileName` attributes and for the regenerated PDB.

Reply:
Not sure what you're getting at here. This document is just applicable to the validator host to download documents to the machine. Hence the only info they should need is enough info to get the files onto the local machine. All the info about file naming should still be within the `dotnet-csharp-validator` tool itself.

Reply:
How will the validator find the source files though?

Reply:
The manifest includes the name of the file that the host should use when putting the file on disk. Hence we can use whatever name we want for htat name. My intent was to just use the hash of the file as the "name". That makes it very easy to line up the document names with the content on disk since the hash is avaiable with the document in the PDB. 

More plainly stated. For a document like `src/Compilers/Shared.cs` we add the output name in the manifest to be whatever the PDB says in the checksum of that file. Hence when re-constructing the compilation we enumerate the document table in the PDB, grab the hash and look for that file name in the output folder. That should effectively avoid collisions and allow for a straight forward mapping.

Reply:
Oh, I see. You'll find the name by the checksum in the PDB. Makes sense.